utils/bundle_miner.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2019 IOTA Stiftung
+ * https://github.com/iotaledger/entangled
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+
+#include <math.h>
+#include <stdbool.h>
+
+#include "common/crypto/iss/normalize.h"
+#include "common/crypto/kerl/kerl.h"
+#include "common/defs.h"
+#include "common/trinary/trit_long.h"
+#include "utils/bundle_miner.h"
+
+#define OBSOLETE_TAG_OFFSET (243 + 81)
+#define OBSOLETE_TAG_LENGTH 81
+
+/**
+ * Public functions
+ */
+
+double bundle_miner_probability_of_losing(byte_t const *const nb, uint64_t const n) {
+  double probability = 0.0;
+  double pi = 0.0;
+
+  for (size_t i = 0; i < n * NORMALIZED_FRAGMENT_LENGTH; i++) {
+    pi = 1.0 - ((double)(TRYTE_VALUE_MAX - nb[i]) / (double)(TRYTE_VALUE_MAX - TRYTE_VALUE_MIN));
+    if (pi > 0.0) {
+      if (probability == 0.0) {
+        probability = 1.0;
+      }
+      probability *= pi;
+    }
+  }
+
+  return probability;
+}
+
+double bundle_miner_security_level(double const probability, double const radix) {
+  return log(1.0 / probability) / log(radix);
+}
+
+void bundle_miner_min_normalized_bundle(byte_t const *const a, byte_t const *const b, byte_t *const out,
```

Comment:
Another option would be to calculate a score for each bundle which is the sum of the values of all trytes and the one with maximal sum get to be chosen, calling this function does not guarantee that the new bundle won't contain a tryte which its value is not related to an already revealed key fragment if that was the original intention,
assume a bundle with only two trytes _min =_ `BA` and a new mined bundle _nb =_ `AZ`, so `out` will be `BZ` which will get a very good score from `bundle_miner_probability_of_losing` so this means _nb_ will be chosen which has `A` with a lower tryte value, meaning an exposed key fragment,
this is ok, you can't expect to mine a bundle without any exposed fragments, but this method is not doing what i suspect it intended to do, plus, what does _out_ is not a real bundle so we should not calculate a score based on it,
we should instead calculate a score for each bundle we mine and the reference one and choose the best